### ISSUE TITLE
Can Inverted be added to the Switch item?

### ISSUE THREAD

**Issue by tonyyng** [openhab/openhab-google-assistant#149]
I'm trying to link my Tesla trunk to a Google Assistant action and I'm finding that the link is reversed.
Here is my item:

`Switch Tesla_RearTrunk "Tesla Trunk" {channel="tesla:model3:myaccount:mycar:reartrunk", ga="Lock"}
`
Command: "Okay Google, Unlock Tesla Trunk"
Response: "Okay, unlocking Tesla Trunk"
Google Binding: sends OFF
Tesla Binding: OFF is used to close the trunk. (WRONG) 

If I use a ga type of Switch, it works, but it isn't ideal:

`Switch Tesla_RearTrunk "Tesla Trunk" {channel="tesla:model3:myaccount:mycar:reartrunk", ga="Switch"}
`
Command: "Okay Google, open Tesla Trunk"  **_(Google doesn't understand "unlock" with switches)_**
Response: "Okay, **_turning on_** Tesla Trunk" **_(Turning on the trunk?)_**
Google Binding: sends ON
Tesla Binding: ON opens the trunk

Can Inverted be supported for Switches so I can do this?
`Switch Tesla_RearTrunk "Tesla Trunk" {channel="tesla:model3:myaccount:mycar:reartrunk", ga="Lock" [Inverted=true]}
`
or is there another way to approach this?

Here is a [link](https://www.openhab.org/addons/bindings/tesla/#channels) to the Tesla binding addon.

**Comment by michikrug** [comment]
I think this will be easy to implement.

Have you thought about using the "Door" type with the Switch item?

**Comment by tonyyng** [comment]
Yes, I tried that, but the Google Assistant Door type only works with a RollerShutter item.  The Tesla binding exposes the reartrunk and fronttrunk as Switches.

**Comment by michikrug** [comment]
Ah, I see

**Comment by michikrug** [comment]
I might extend those types to be also used as switches

**Comment by tonyyng** [comment]
That would be great.

**Comment by michikrug** [comment]
Added a PR for that.

**Comment by michikrug** [comment]
Should be available now

**Comment by sgtstadanko** [comment]
I just did a fresh latest pull from docker image and Inverted=true isnt working for these Tesla switches. @tonyyng were you able to get it working?

**Comment by tonyyng** [comment]
I updated my docker container to openhab:2.5.4 to test it, and the behavior is the same.  With or without inverted, the behavior is backwards.

`Switch M3_RearTrunk "Trunk" {channel="tesla:model3:myaccount:mycar:reartrunk", ga="Lock" [Inverted=true] }`
Restart Openhab
`2020-05-07 14:44:14.103 [vent.ItemStateChangedEvent] - M3_DoorLock changed from NULL to ON`
`2020-05-07 14:44:14.106 [vent.ItemStateChangedEvent] - M3_RearTrunk changed from NULL to OFF`

With the trunk closed, I say "Okay, Google, lock the trunk".  (Trunk opens)
`2020-05-07 14:47:49.462 [ome.event.ItemCommandEvent] - Item 'M3_RearTrunk' received command ON`
`2020-05-07 14:47:49.464 [nt.ItemStatePredictedEvent] - M3_RearTrunk predicted to become ON`
`2020-05-07 14:47:49.476 [vent.ItemStateChangedEvent] - M3_RearTrunk changed from OFF to ON`

Manually close the trunk
`2020-05-07 14:47:51.293 [vent.ItemStateChangedEvent] - M3_RearTrunk changed from ON to OFF`
`2020-05-07 14:47:58.269 [vent.ItemStateChangedEvent] - M3_RearTrunk changed from OFF to NULL`
`2020-05-07 14:48:12.978 [vent.ItemStateChangedEvent] - M3_RearTrunk changed from NULL to OFF`

I tried it again, removing [Inverted=true], restarting openhab, and the behavior (including logs) is exactly the same.

Do I have to use one of the experimental snapshot versions to get the change?   Is there a way to confirm I'm running the correct version of the Google Assistant code?

**Comment by michikrug** [comment]
First of all, the Google Assistant integration is not bound to any openHAB release. So updating openHAB or restarting it will not have any influence in the behavior.
Second, `inverted` needs to be lowercase as the documentation states.
Third, you need to sync your devices with GA to populate changes on items in openHAB.

**Comment by tonyyng** [comment]
I just tried it with the correct case:

`Switch M3_RearTrunk "Trunk" {channel="tesla:model3:myaccount:mycar:reartrunk"`
`                    ,ga="Lock" [inverted=true]`
`                    }`

Restarted openhab, asked Google to resync "Hey Google, resync openhab", and the behavior is still the same.

Is there a way to confirm that the item has the inverted attribute set at runtime?  I looked in the rules documentation, and didn't come across anything.

**Comment by sgtstadanko** [comment]
Was the change merged into the Google assistant integration  code or the
main Openhab?  If its separate then updating openhab won't do anything.

Also if you are running with the openhab cloud integration and myopenhab,
you don't have to sync or represent. It just needs an item file with
correct Metadata or tag.

If you are using Google assistant the easiest way around this is to add a
routine that listens for "ok Google, pop the trunk" and then bind the
action to "close trunk" or whatever you named your switch in the item file.

**Comment by tonyyng** [comment]
I'm running with myopenhab and the cloud with what I think is the correct item configuration (now)

I saw your suggested workaround.  I see how that could be used to make it work.  You'll still get the wrong answer if you ask "Hey Google, is the trunk open?"

To be honest, it isn't that important a feature that I'll worry too much about getting it working.  I have the Tesla app on my phone to do most what I need.

However, since Michael added this feature at my request, I feel obliged to help with the debugging.  It's very likely I did something wrong with my configuration.

**Comment by michikrug** [comment]
Ok, sorry guys. I just had a look at the code.
The inverted option is only available for switches and rollershutters.
`Lock` is handled separately, and therefore the inverted option is not available.
You will need to use e.g. `Door`.

**Comment by tonyyng** [comment]
Just tried Door with no success.  Is it because the ga="Door" only works with Rollershutter?

Will Door work with a Switch item type?

**Comment by michikrug** [comment]
Yes, this is what I implemented recently

**Comment by tonyyng** [comment]
@michikrug Based on your explanation, the GA integration code "works completely standalone paired with openHAB Cloud."

The log messages I've been pasting are from openhab running on my raspberry pi.  

I shouldn't expect them to change based on the item's inverted setting, right?

If the trunk is closed, I would expect to see:
`2020-05-07 14:44:14.106 [vent.ItemStateChangedEvent] - M3_RearTrunk changed from NULL to OFF`
at startup regardless of whether inverted was true or false, correct?

The difference would be in the command received from the Google Assistant integration
With the trunk closed, the voice command "Okay, Google, lock the trunk" would result in
`Item 'M3_RearTrunk' received command OFF` if inverted = true and
`Item 'M3_RearTrunk' received command ON` if inverted = false.  Am I understanding it correctly?

> Yes, this is what I implemented recently
Has this change been deployed?  I'll try ga="Door" tomorrow.

**Comment by michikrug** [comment]
Yes, this is correct.

And yes, it is deployed as others also reported the inverted works for switches in the openhab forum.

**Comment by tonyyng** [comment]
I got the trunk to open correctly with this item definition:

`Switch M3_RearTrunk "Tesla Trunk" {channel="tesla:model3:myaccount:mycar:reartrunk", ga="Door"}`

Does the [documentation](https://www.openhab.org/docs/ecosystem/google-assistant/) need to change to indicate that  `Switch {ga="Door"}` is a valid combination?

Thanks for your help. michikrug

ps.  It took some experimentation with the name I used for the item before the Google Assistant would execute the command instead of showing me search results.

For example, "Hey Google.  Open the trunk." would result in the assistant showing me search results for "open trunk".

This is a Google Assistant quirk;  nothing to do with Openhab.
